### PR TITLE
Add implementation for Ponos App & Service

### PR DESCRIPTION
**Pull request by spirosoik**
#### Summary
Create a Mattermost Application for ChatOps toil elimination.
Create a Service which interacts with Ponos ChatOps

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41015

### CODE REVIEW

**Comment by spirosoik** [comment]
@stylianosrigas @stafot CI is in process check the logic for now

File: workspaces/service.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		logger:            logger,
+	}
+}
+
+func (s *Service) DeleteWorkspace(dnsName string) error {
+	installation, err := s.provisionerClient.GetInstallationByDNS(dnsName, &cmodel.GetInstallationRequest{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get installation by DNS")
+	}
+	return s.provisionerClient.DeleteInstallation(installation.ID)
```
Comment (stafot) [inline workspaces/service.go]: For the deletion this is not enough as there are needed some extra disassociations that are not handled by the provisioner.

Comment (spirosoik) [reply workspaces/service.go]: Ok we can do more that's what easy

File: workspaces/client.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		client: client,
+	}
+}
+
+// DeleteWorkspace makes an http request to delete a workspace
+func (c *WorkspaceClient) DeleteWorkspace(installationID string) error {
+	if installationID == "" {
+		return errors.New("failed to delete workspace, installationID is required")
+	}
+	endpoint := fmt.Sprintf("%s/installation/%s", c.URL, installationID)
```
Comment (stafot) [inline workspaces/client.go:28]: 👍🏾

File: function/app.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	app.HandleCommand(configureCommand)
+	app.HandleCommand(addUserCommand)
+
+	// Cloud workspace handlers
+	app.HandleCall("/workspaces/delete", goapp.RequireConnectedUsers(goapp.CallHandler(app.deleteWorkspaceHandler)))
+	app.HandleCall("/form/delete-workspace", goapp.RequireConnectedUsers(goapp.FormHandler(app.deleteWorkspaceFormHandler)))
+
+	// Cloud migration handlers
+	app.HandleCall("/migrations/create", goapp.RequireConnectedUsers(goapp.CallHandler(app.createMigrationHandler)))
+	app.HandleCall("/form/delete-workspace", goapp.RequireConnectedUsers(goapp.FormHandler(app.createMigrationFormHandler)))
```
Comment (stylianosrigas) [inline function/app.go]:
```suggestion
	app.HandleCall("/form/migrate-workspace", goapp.RequireConnectedUsers(goapp.FormHandler(app.createMigrationFormHandler)))
```

File: internal/api/ponos_client.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+// - IAM resouces & policy
+func (c *PonosClient) CreateMiration(data migrations.CustomerMigrationRequest) error {
+	body, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse create migration request payload")
+	}
+	endpoint := fmt.Sprintf("%s/migrations", c.url)
+	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(body))
+	if err != nil {
+		return errors.Wrap(err, "failed to create a request to delete a workspace")
```
Comment (stylianosrigas) [inline internal/api/ponos_client.go]:
```suggestion
		return errors.Wrap(err, "failed to create a request to migrate a workspace")
```

File: internal/api/ponos_client.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+// - IAM resouces & policy
+func (c *PonosClient) DeleteMiration(data migrations.CustomerMigrationRequest) error {
+	body, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse create migration request payload")
+	}
+	endpoint := fmt.Sprintf("%s/migrations", c.url)
+	req, err := http.NewRequest("DELETE", endpoint, bytes.NewBuffer(body))
+	if err != nil {
+		return errors.Wrap(err, "failed to create a request to delete a workspace")
```
Comment (stylianosrigas) [inline internal/api/ponos_client.go]:
```suggestion
		return errors.Wrap(err, "failed to create a request to delete a migration")
```

File: internal/api/ponos_client.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// DeleteMiration will delete the created resources for
+// migrating a customer in cloud with AWAT. Resources:
+// - S3 Bucket
+// - IAM resouces & policy
+func (c *PonosClient) DeleteMiration(data migrations.CustomerMigrationRequest) error {
+	body, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse create migration request payload")
```
Comment (stylianosrigas) [inline internal/api/ponos_client.go]:
```suggestion
		return errors.Wrap(err, "failed to parse delete migration request payload")
```

File: internal/api/ponos_client.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	if err != nil {
+		return errors.Wrap(err, "failed to create a request to delete a workspace")
+	}
+	return c.do(req)
+}
+
+func (c *PonosClient) do(req *http.Request) error {
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete a workspace")
```
Comment (stylianosrigas) [inline internal/api/ponos_client.go]:
```suggestion
		return errors.Wrap(err, "failed to execute a ponos client command")
```

File: migrations/service.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	if err := tf.Plan(dto.toTerraformArgs()...); err != nil {
+		return errors.Wrap(err, "failed to run Terraform plan")
+	}
+	s.logger.Info("successfully ran Terraform plan")
+
+	if dto.Apply {
+		s.logger.Info("destroying migration infra")
+		if err := tf.Destroy(dto.toTerraformArgs()...); err != nil {
+			return errors.Wrap(err, "failed to apply migration infra")
```
Comment (stylianosrigas) [inline migrations/service.go]:
```suggestion
			return errors.Wrap(err, "failed to destroy migration infra")
```

File: workspaces/client.go (hunk 2 above)
Comment (stylianosrigas) [reply workspaces/client.go:28]: @spirosoik does it handle the customer web server deletion too?

**Review by stylianosrigas — CHANGES_REQUESTED** [review]
Just some nit comments. Amazing work Spiros ❤️ 🚀

File: internal/api/ponos_client.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	if resp.StatusCode > 400 && resp.StatusCode < 500 {
```
Comment (angeloskyratzakos) [inline internal/api/ponos_client.go]: Not to lose 400 code
```suggestion
	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
```

File: workspaces/dto.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+// DeleteWorkspaceRequest the data transfer object to delete a
+// a workspace with the provided DNS name
+type DeleteWorkspaceRequest struct {
+	Name string `json:"dns_name"`
+}
+
+// Validate validates the values of a delete workspace request
+func (r *DeleteWorkspaceRequest) Validate() error {
+	if r.Name == "" {
+		return errors.New("DNS name cannot be emp")
```
Comment (angeloskyratzakos) [inline workspaces/dto.go]: nit:
```suggestion
		return errors.New("DNS name cannot be empty")
```

File: workspaces/dto.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package workspaces
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+// DeleteWorkspaceRequest the data transfer object to delete a
```
Comment (angeloskyratzakos) [inline workspaces/dto.go]: nit:
```suggestion
// DeleteWorkspaceRequest the data transfer object to delete
```

File: workspaces/client.go (hunk 2 above)
Comment (spirosoik) [reply workspaces/client.go:28]: yes actually this is the part here.

File: workspaces/client.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+	req, err := http.NewRequest("DELETE", endpoint, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to create a delete HTTP request")
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete a workspace")
+	}
+	if resp.StatusCode > 400 && resp.StatusCode < 500 {
```
Comment (angeloskyratzakos) [inline workspaces/client.go]:
```suggestion
	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
```

File: LICENSE [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
Comment (angeloskyratzakos) [inline LICENSE]: I think we need to update this one to be like the boilerplate
```suggestion
   Copyright 2015-present Mattermost, Inc.
```

File: internal/api/ponos_client.go (hunk 5 above)
Comment (stafot) [reply internal/api/ponos_client.go]: @stylianosrigas is there a typo on delete wording?

**Comment by stafot** [comment]
vvv minor but let's use main instead of master since is a newly created repository.

**Review by stylianosrigas — APPROVED** [review]
LGTM 👌

**Review by angeloskyratzakos — APPROVED** [review]

**Review by stafot — APPROVED** [review]

**Pull request merged**